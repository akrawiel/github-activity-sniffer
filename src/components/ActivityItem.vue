<template lang="pug">
transition(name="collapse" appear)
  .activity-item
    img.avatar(:src="member.avatar_url")
    .member-login.flex.font-bold.items-center.justify-start.text-xl 
      a.link(:href="memberProfileHref" target="_blank") {{ member.login }}
    .latest-event-container
      .latest-event-date.gap-2.grid.grid-columns-manual.justify-between.w-full
        b Latest event date:&emsp;
        span.truncate.text-right {{ latestEventDate }}
      .latest-event-type.gap-2.grid.grid-columns-manual.justify-between.w-full
        b Latest event type:&emsp;
        span.truncate.text-right {{ latestEventType }}
      .latest-event-repo.gap-2.grid.grid-columns-manual.justify-between.w-full
        b Latest event repo:&emsp;
        a.link.truncate.text-right(:href="latestEventRepoHref" target="_blank" v-if="latestEventRepoHref") {{ latestEventRepoName }}
        span.truncate.text-right(v-else) {{ latestEventRepoName }}
</template>

<script>
import { format } from "date-fns";

export default {
  props: {
    member: Object
  },
  computed: {
    latestEventDate() {
      const createdAt = this.member?.latest_event?.created_at ?? null;

      return createdAt
        ? format(new Date(createdAt), "yyyy-MM-dd HH:mm:ss")
        : "N/A";
    },
    latestEventType() {
      return this.member?.latest_event?.type ?? "N/A";
    },
    latestEventRepoHref() {
      const repoName = this.member?.latest_event?.repo?.name;
      return repoName ? `https://github.com/${repoName}` : null;
    },
    latestEventRepoName() {
      return this.member.latest_event?.repo?.name ?? "N/A";
    },
    memberProfileHref() {
      return `https://github.com/${this.member.login}`;
    }
  }
};
</script>

<style lang="postcss" scoped>
$avatar_size: 5rem;
$item_height: 6.25rem;

.activity-item {
  @apply grid gap-4 items-center overflow-hidden p-2 w-full;
  grid-template-columns: $avatar_size 0.5fr 0.5fr;
  height: $item_height;

  @media screen and (max-width: theme("screens.sm")) {
    grid-template-columns: 1fr;
  }
}

.avatar {
  @apply border-4 border-solid border-blue-400 rounded-full shadow-lg;
  height: $avatar_size;
  width: $avatar_size;
}

.collapse-enter,
.collapse-leave-to {
  height: 0;
}

.collapse-enter-to,
.collapse-leave {
  height: $item_height;
}

.collapse-enter-active {
  transition: height 0.25s ease-out;
}

.latest-event-container {
  @apply flex flex-col items-end justify-center text-xs;

  @media screen and (max-width: theme("screens.sm")) {
    @apply items-start;
  }
}

.link {
  @apply text-blue-700;
  transition: color 0.125s ease-out;
  &:hover {
    @apply text-blue-400;
  }
}

.grid-columns-manual {
  grid-template-columns: auto 1fr;
}
</style>
