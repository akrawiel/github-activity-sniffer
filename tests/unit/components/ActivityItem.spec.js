import { shallowMount } from "@vue/test-utils";
import ActivityItem from "@/components/ActivityItem.vue";
import { format } from "date-fns";

describe("ActivityItem.vue", () => {
  it("renders when all data is passed", () => {
    const propsData = {
      member: {
        avatar_url: "Avatar",
        login: "Login",
        latest_event: {
          created_at: "2020-02-02T01:00:00Z",
          type: "Test",
          repo: {
            name: "test/test"
          }
        }
      }
    };

    const wrapper = shallowMount(ActivityItem, {
      propsData
    });

    expect(wrapper).toMatchSnapshot();

    expect(wrapper.find(".avatar").attributes("src")).toMatch("Avatar");

    expect(wrapper.find(".member-login").text()).toMatch("Login");

    expect(
      wrapper
        .find(".latest-event-date")
        .text()
        .replace(/\s/g, " ")
    ).toMatch(
      `Latest event date: ${format(
        new Date(propsData.member.latest_event.created_at),
        "yyyy-MM-dd HH:mm:ss"
      )}`
    );

    expect(wrapper.find(".latest-event-type").text()).toMatch("Test");

    const repoLink = wrapper.find(".latest-event-repo").find(".link");

    expect(repoLink.attributes("href")).toMatch("https://github.com/test/test");

    expect(repoLink.text()).toMatch("test/test");
  });

  it("renders when latest event is missing", () => {
    const propsData = {
      member: {
        avatar_url: "Avatar",
        login: "Login"
      }
    };

    const wrapper = shallowMount(ActivityItem, {
      propsData
    });

    expect(wrapper).toMatchSnapshot();

    expect(wrapper.find(".avatar").attributes("src")).toMatch("Avatar");

    expect(wrapper.find(".member-login").text()).toMatch("Login");

    expect(
      wrapper
        .find(".latest-event-date")
        .text()
        .replace(/\s/g, " ")
    ).toMatch(`Latest event date: N/A`);

    expect(wrapper.find(".latest-event-type").text()).toMatch("N/A");

    const repoLink = wrapper.find(".latest-event-repo").find(".link");

    expect(repoLink.exists()).toBeFalsy();
  });
});
