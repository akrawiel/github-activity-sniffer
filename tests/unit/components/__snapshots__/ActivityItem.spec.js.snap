// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityItem.vue renders when all data is passed 1`] = `
<div
  class="activity-item"
>
  <img
    class="avatar"
    src="Avatar"
  />
  <div
    class="member-login flex font-bold items-center justify-start text-xl"
  >
    <a
      class="link"
      href="https://github.com/Login"
      target="_blank"
    >
      Login
    </a>
  </div>
  <div
    class="latest-event-container"
  >
    <div
      class="latest-event-date gap-2 grid grid-columns-manual justify-between w-full"
    >
      <b>
        Latest event date: 
      </b>
      <span
        class="truncate text-right"
      >
        2020-02-02 02:00:00
      </span>
    </div>
    <div
      class="latest-event-type gap-2 grid grid-columns-manual justify-between w-full"
    >
      <b>
        Latest event type: 
      </b>
      <span
        class="truncate text-right"
      >
        Test
      </span>
    </div>
    <div
      class="latest-event-repo gap-2 grid grid-columns-manual justify-between w-full"
    >
      <b>
        Latest event repo: 
      </b>
      <a
        class="link truncate text-right"
        href="https://github.com/test/test"
        target="_blank"
      >
        test/test
      </a>
    </div>
  </div>
</div>
`;

exports[`ActivityItem.vue renders when latest event is missing 1`] = `
<div
  class="activity-item"
>
  <img
    class="avatar"
    src="Avatar"
  />
  <div
    class="member-login flex font-bold items-center justify-start text-xl"
  >
    <a
      class="link"
      href="https://github.com/Login"
      target="_blank"
    >
      Login
    </a>
  </div>
  <div
    class="latest-event-container"
  >
    <div
      class="latest-event-date gap-2 grid grid-columns-manual justify-between w-full"
    >
      <b>
        Latest event date: 
      </b>
      <span
        class="truncate text-right"
      >
        N/A
      </span>
    </div>
    <div
      class="latest-event-type gap-2 grid grid-columns-manual justify-between w-full"
    >
      <b>
        Latest event type: 
      </b>
      <span
        class="truncate text-right"
      >
        N/A
      </span>
    </div>
    <div
      class="latest-event-repo gap-2 grid grid-columns-manual justify-between w-full"
    >
      <b>
        Latest event repo: 
      </b>
      <span
        class="truncate text-right"
      >
        N/A
      </span>
    </div>
  </div>
</div>
`;
